src/AnnotationReader.php

```diff
@@ -0,0 +1,279 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getPropertyAnnotations(ReflectionProperty $property) : array
+    {
+        $class     = $property->getDeclaringClass();
+        $docblock  = $property->getDocComment();
+        $namespace = $class->getNamespaceName();
+        $imports   = $this->getPropertyImports($property);
+        $ignored   = $this->getIgnoredAnnotationNames($class);
+        $context   = new Context($property, [$namespace], $imports, $ignored);
+
+        if ($docblock === false) {
+            return [];
+        }
+
+        return $this->docParser->parse($docblock, $context);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getPropertyAnnotation(ReflectionProperty $property, string $annotationName)
+    {
+        $annotations = $this->getPropertyAnnotations($property);
+
+        foreach ($annotations as $annotation) {
+            if ($annotation instanceof $annotationName) {
+                return $annotation;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getMethodAnnotations(ReflectionMethod $method) : array
+    {
+        $class     = $method->getDeclaringClass();
+        $docblock  = $method->getDocComment();
+        $namespace = $class->getNamespaceName();
+        $imports   = $this->getMethodImports($method);
+        $ignored   = $this->getIgnoredAnnotationNames($class);
+        $context   = new Context($method, [$namespace], $imports, $ignored);
+
+        if ($docblock === false) {
+            return [];
+        }
+
+        return $this->docParser->parse($docblock, $context);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getMethodAnnotation(ReflectionMethod $method, string $annotationName)
+    {
+        $annotations = $this->getMethodAnnotations($method);
+
+        foreach ($annotations as $annotation) {
+            if ($annotation instanceof $annotationName) {
+                return $annotation;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the ignored annotations for the given class.
+     *
+     * @param \ReflectionClass $class
+     *
+     * @return array
+     */
+    private function getIgnoredAnnotationNames(ReflectionClass $class) : array
+    {
+        $ignoredNames = $this->config->getIgnoredAnnotationNames()->getArrayCopy();
+        $annotations  = $this->metadataParser->parseAnnotationClass($class);
+
+        foreach ($annotations as $annotation) {
+            if ( ! $annotation instanceof IgnoreAnnotation) {
+                continue;
+            }
+
+            foreach ($annotation->names as $name) {
+                $ignoredNames[$name] = true;
+            }
+        }
+
+
+        return $ignoredNames;
+    }
+
+    /**
+     * Retrieves imports.
+     *
+     * @param \ReflectionClass $class
+     *
+     * @return array
+     */
+    private function getClassImports(ReflectionClass $class)
```

Comment:
These methods should be moved to a separate class, responsible to handle ReflectionClass specifics that could potentially disappear if we have better PHP support (such as extracting "use"s). I'd suggest ReflectionClass, ReflectionMethod and ReflectionProperty under a Reflection namespace. Also we'd require a ReflectionClassFactory that could keep the cached imports of a given class.

Example:
```php
$this->reflectionClassFactory = new Reflection\ReflectionClassFactory();

$reflectionClass = this->reflectionClassFactory->getClass($reflClass); // instanceof Reflection\ReflectionClass
$reflectionProperty = $reflectionClass->getProperty($reflProperty); // instanceof Reflection\ReflectionProperty
$reflectionMethod = $reflectionClass->getMethod($reflMethod); // instanceof Reflection\ReflectionMethod

$reflectionClass->getImports()
$reflectionProperty->getImports()
$reflectionMethod->getImports()
```